clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -109,78 +112,76 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 // we have enough unallocated or pooled memory to immediately
                 // satisfy the request
                 freeUp(size);
+                ByteBuffer allocatedBuffer = allocateByteBuffer(size);
```

Comment:
I think it was intentional that we didn't allocate memory with the lock held. It seems like this optimisation was lost?